### ISSUE TITLE
The parser cannot parse non-normal values!

### ISSUE THREAD

**Issue by ChevronOne** [Tencent/rapidjson#2099]
Hi, 
The parser `GenericDocument::Parse(const Ch* str)` fails to parse non-normal values such as `INFINITY`, `-INFINITY` and `NAN` for floating point data types. These exceptional values are supported officially by the standard, however the parser still can't parse them correctly!

This is actually very argent from our end, thus any soon feedback would be appreciated;-) thanks.

**Comment by ChevronOne** [comment]
Update: 
We just figured out that there is indeed a flag (`kParseNanAndInfFlag`) that can be given to allow parsing these values.
The remaining problem however is that we also find out in the source code that these values are being tested against `Inf`, `-Inf` and `NaN`, but according to the standard, e.g. when using `std::to_string( double )` , these values get converted as `inf`, `-inf`, `nan`, therefore the parse unfortunately still fails to parse them!

**Comment by miloyip** [comment]
Shall this be relaxed to case-insensitve?

**Comment by ChevronOne** [comment]
Yes, that's what I thought too at the first glance. However, it wasn't that intuitive, as the `nan` was then getting mis-parsed with `null` as both start with a lowercase `n` but being parsed in different function. I just provided a fix in this [pull request](https://github.com/Tencent/rapidjson/pull/2100), so you may have a look.
